tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -718,11 +727,24 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     // a matrix bias is only supported with CUDA 12 and above.
     HloInstruction *c = nullptr;
 #if CUDA_VERSION > 12000
+    auto compatible_c_type = [&](const PrimitiveType ctype) {
+      PrimitiveType dtype = instr->shape().element_type();
+      if (ctype == BF16) {
+        return dtype == F8E4M3FN || dtype == F8E5M2 || dtype == BF16;
+      } else if (ctype == F16) {
+        return dtype == F8E4M3FN || dtype == F8E5M2 || dtype == F16;
+      } else if (ctype == F32) {
+        return dtype == F32;
+      }
+      return false;
+    };
+
     if (instr->user_count() == 1 &&
         instr->users()[0]->opcode() == HloOpcode::kAdd) {
       HloInstruction *add = instr->users()[0];
       HloInstruction *bias = add->mutable_operand(!add->operand_index(instr));
-      if (bias->opcode() != HloOpcode::kBroadcast) {
+      if (bias->opcode() != HloOpcode::kBroadcast &&
+          compatible_c_type(bias->shape().element_type())) {
```

Comment:
the type of `bias` and the type of `instr` are guaranteed to be the same, since the Add instruction's inputs and output must all be the same type. So this check is unnecessary.

Reply:
Right. Will remove.